crates/bevy_ecs/src/system/function_system.rs

```diff
@@ -490,6 +482,20 @@ where
         ));
     }
 
+    fn update_archetypes(&mut self, world: &World) {
+        let archetypes = world.archetypes();
+        let new_generation = archetypes.generation();
+        let old_generation = std::mem::replace(&mut self.archetype_generation, new_generation);
+        let archetype_index_range = old_generation.value()..new_generation.value();
+
+        for archetype_index in archetype_index_range {
+            self.param_state.as_mut().unwrap().new_archetype(
+                &archetypes[ArchetypeId::new(archetype_index)],
+                &mut self.system_meta,
+            );
+        }
+    }
```

Comment:
I think you should also store a `WorldId` and confirm that it matches here, just so that there's full parity between `FunctionSystem` and `SystemState` (would help with the lifecycle rework).

Reply:
Added.